core/src/main/java/com/linecorp/armeria/client/retry/RetryConfig.java

```diff
@@ -0,0 +1,224 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(RetryConfig.class);
+
+    /**
+     * Returns a new {@link RetryConfigBuilder} with the default values from Flags.
+     * Uses a {@code RetryRule}.
+     */
+    public static RetryConfigBuilder<HttpResponse> builder(RetryRule retryRule) {
+        return new RetryConfigBuilder<>(retryRule);
+    }
+
+    /**
+     * Returns a new {@link RetryConfigBuilder} with the default values from Flags.
+     * Uses a {@code RetryRuleWithContent<HttpResponse>}.
+     */
+    public static RetryConfigBuilder<HttpResponse> builder(
+            RetryRuleWithContent<HttpResponse> retryRuleWithContent) {
+        return new RetryConfigBuilder<>(retryRuleWithContent);
+    }
+
+    /**
+     * Returns a new {@link RetryConfigBuilder} with the default values from Flags.
+     * Uses a {@code RetryRule}.
+     */
+    public static RetryConfigBuilder<RpcResponse> builderForRpc(RetryRule retryRule) {
+        return new RetryConfigBuilder<>(retryRule);
+    }
+
+    /**
+     * Returns a new {@link RetryConfigBuilder} with the default values from Flags.
+     * Uses a {@code RetryRuleWithContent<RpcResponse>}.
+     */
+    public static RetryConfigBuilder<RpcResponse> builderForRpc(
+            RetryRuleWithContent<RpcResponse> retryRuleWithContent) {
+        return new RetryConfigBuilder<>(retryRuleWithContent);
+    }
+
+    static <T extends Response> RetryConfigBuilder<T> builder0(RetryRule retryRule) {
+        return new RetryConfigBuilder<>(retryRule);
+    }
+
+    static <T extends Response> RetryConfigBuilder<T> builder0(
+            RetryRuleWithContent<T> retryRuleWithContent) {
+        return new RetryConfigBuilder<>(retryRuleWithContent);
+    }
+
+    private final int maxTotalAttempts;
+    private final long responseTimeoutMillisForEachAttempt;
+    private final int maxContentLength;
+
+    @Nullable
+    private final RetryRule retryRule;
+
+    @Nullable
+    private final RetryRuleWithContent<T> retryRuleWithContent;
+
+    @Nullable
+    private final RetryRule fromRetryRuleWithContent;
+
+    @Nullable
+    private RetryRuleWithContent<T> fromRetryRule;
+
+    RetryConfig(RetryRule retryRule, int maxTotalAttempts, long responseTimeoutMillisForEachAttempt) {
+        checkArguments(maxTotalAttempts, responseTimeoutMillisForEachAttempt);
+        this.retryRule = requireNonNull(retryRule, "retryRule");
+        this.maxTotalAttempts = maxTotalAttempts;
+        this.responseTimeoutMillisForEachAttempt = responseTimeoutMillisForEachAttempt;
+        maxContentLength = 0;
+        retryRuleWithContent = null;
+        fromRetryRuleWithContent = null;
+    }
+
+    RetryConfig(
+            RetryRuleWithContent<T> retryRuleWithContent,
+            int maxContentLength,
+            int maxTotalAttempts,
+            long responseTimeoutMillisForEachAttempt) {
+        checkArguments(maxTotalAttempts, responseTimeoutMillisForEachAttempt);
```

Comment:
How about making a private constructor and check it there?
```java
RetryConfig(RetryRule retryRule, int maxTotalAttempts, long responseTimeoutMillisForEachAttempt) {
    this(requireNonNull(retryRule, "retryRule"),
         null, 0, maxTotalAttempts, responseTimeoutMillisForEachAttempt);
}

RetryConfig(RetryRuleWithContent<T> retryRuleWithContent, int maxContentLength, int maxTotalAttempts,
            long responseTimeoutMillisForEachAttempt) {
    this(null, requireNonNull(retryRuleWithContent, "retryRuleWithContent"), maxContentLength,
         maxTotalAttempts, responseTimeoutMillisForEachAttempt);
}

private RetryConfig(@Nullable RetryRule retryRule, @Nullable RetryRuleWithContent<T> retryRuleWithContent,
                    int maxContentLength, int maxTotalAttempts, long responseTimeoutMillisForEachAttempt) {
    checkArgument(maxTotalAttempts > 0, "maxTotalAttempts: %s (expected: > 0)", maxTotalAttempts);
    ...
}
```

Reply:
Done